src/devices/sound/qsoundhle.h

```diff
@@ -34,23 +35,133 @@ class qsound_hle_device : public device_t, public device_sound_interface, public
 	virtual void rom_bank_updated() override;
 
 private:
+
+	// DSP ROM sample map
+	enum {
+		DATA_PAN_TAB		= 0x110,
+		STATE_INIT1			= 0x288,
+		STATE_INIT2			= 0x61a,
+		STATE_REFRESH1		= 0x039,
+		STATE_REFRESH2		= 0x04f,
+		STATE_NORMAL1		= 0x314,
+		STATE_NORMAL2 		= 0x6b2,
+		DATA_ADPCM_TAB		= 0x9dc,
+		DATA_FILTER_TAB		= 0xd53,	// dual filter mode, 5 tables * 95 taps each
+		DATA_FILTER_TAB2	= 0xf2e		// overlapping data (95+15+95)
+	};
+
+	const uint16_t PAN_TABLE_DRY = 0;
+	const uint16_t PAN_TABLE_WET = 98;
+	const uint16_t PAN_TABLE_CH_OFFSET = 196;
+	const uint16_t FILTER_ENTRY_SIZE = 95;
+	const uint16_t DELAY_BASE_OFFSET = 0x554;
+	const uint16_t DELAY_BASE_OFFSET2 = 0x53c;
+
+	struct qsound_voice {
+		uint16_t bank;
+		int16_t addr; // top word is the sample address
+		uint16_t phase;
+		uint16_t rate;
+		int16_t loop_len;
+		int16_t end_addr;
+		int16_t volume;
+		int16_t echo;
+	};
+
+	struct qsound_adpcm {
+		uint16_t start_addr;
+		uint16_t end_addr;
+		uint16_t bank;
+		int16_t volume;
+		uint16_t flag;
+		int16_t cur_vol;
+		int16_t step_size;
+		uint16_t cur_addr;
+	};
+
+	// Q1 Filter
+	struct qsound_fir {
+		int tap_count;	// usually 95
+		int delay_pos;
+		uint16_t table_pos;
+		int16_t taps[95];
+		int16_t delay_line[95];
+	};
+
+	// Delay line
+	struct qsound_delay {
+		int16_t delay;
+		int16_t volume;
+		int16_t write_pos;
+		int16_t read_pos;
+		int16_t delay_line[51];
+	};
+
+	struct qsound_echo {
+		uint16_t end_pos;
+
+		int16_t feedback;
+		int16_t length;
+		int16_t last_sample;
+		int16_t delay_line[1024];
+		int16_t delay_pos;
+	};
+
 	// MAME resources
 	sound_stream *m_stream;
+	required_region_ptr<uint16_t> m_dsp_rom;
+
+	uint16_t m_data_latch;
+	int16_t m_out[2];
+
+	qsound_voice m_voice[16];
+	qsound_adpcm m_adpcm[3];
+
+	uint16_t m_voice_pan[16+3];
+	int16_t m_voice_output[16+3];
+
+	qsound_echo m_echo;
+
+	qsound_fir m_filter[2];
+	qsound_fir m_alt_filter[2];
+
+	qsound_delay m_wet[2];
+	qsound_delay m_dry[2];
+
+	uint16_t m_state;
+	uint16_t m_next_state;
+
+	uint16_t m_delay_update;
+
+	int m_state_counter;
+	int m_ready_flag;
+
+	uint16_t *register_map[256];
+
+	inline uint16_t read_dsp_rom(uint16_t addr) { return m_dsp_rom[addr&0xfff]; }
+
+	void write_data(uint8_t addr, uint16_t data);
+	uint16_t read_data(uint8_t addr);
+
+	void init_register_map();
+	void update_sample();
 
-	struct qsound_channel
-	{
-		u16 reg[8];    // channel control registers
+	// DSP states
+	void state_init();
+	void state_refresh_filter_1();
+	void state_refresh_filter_2();
+	void state_normal_update();
 
-		// work variables
-		int lvol;           // left volume
-		int rvol;           // right volume
-	} m_channel[16];
+	// sub functions
+	inline int16_t get_sample(uint16_t bank,uint16_t address);
 
-	int m_pan_table[33];    // pan volume table
-	u16 m_data;        // register latch data
+	inline int16_t pcm_update(struct qsound_voice *v, int32_t *echo_out);
```

Comment:
C++ doesn't use C-style "structure tags".  Please don't add unnecessary `struct` keyword in places like this.  Use references rather than pointers where it doesn't make sense for the argument to be null, and it isn't an array.